### PR TITLE
TE-3485: P&S new structure

### PR DESCRIPTION
**Pull request by alexanderM91**
- Developer(s): @alexanderM91 

- Ticket: https://spryker.atlassian.net/browse/TE-3485

- Suite nonsplit PR: spryker/suite-nonsplit#2194

- CORE PR: https://github.com/spryker/spryker/pull/5281

- EventBehavior: https://github.com/spryker/event-behavior/pull/48

- Release Group: https://release.spryker.com/release-groups/view/1414

- Release Overview: https://release.spryker.com/release/pull-request/4460

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   RabbitMq       | minor                 |                         |

#### Release Notes

-----------------------------------------

### Improvements

- Adjusted RabbitMQ configuration so queue and exchange creation are handled by default.

### CODE REVIEW

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php [hunk 1]
```diff
@@ -75,34 +106,89 @@ protected function getQueueConnectionConfigs(): array
...
+            return $this->queueOptionCollection;
+        }
 
-        return $queueOptionCollection;
+        $queueConfigurations = $this->getQueueConfiguration();
+        $this->queueOptionCollection = new ArrayObject();
+
+        foreach ($queueConfigurations as $queueNameKey => $queueConfiguration) {
+            if (!is_array($queueConfiguration)) {
+                $this->queueOptionCollection->append($this->createQueueOption($queueConfiguration, sprintf('%s.error', $queueConfiguration), 'error'));
```
Comment (ehsanmx) [inline src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: It's a bit scary that the core module create OOTB `error` queues, I expect the `error` or any other namings come form project not core

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php [hunk 2]
```diff
@@ -75,34 +106,89 @@ protected function getQueueConnectionConfigs(): array
...
+
+        return $queueOptionTransfer;
+    }
+
+    /**
+     * @param string $queueName
+     *
+     * @return \Generated\Shared\Transfer\RabbitMqOptionTransfer
+     */
+    protected function createQueueBinding($queueName)
```
Comment (ehsanmx) [inline src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: this method is exactly like the `createErrorQueueBinding` but just with an empty `routingKey` maybe you can just merge these two methods together and please drop `error` from any naming in core

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php [hunk 3]
```diff
@@ -75,34 +106,89 @@ protected function getQueueConnectionConfigs(): array
...
+        return $queueOptionTransfer;
+    }
+
+    /**
+     * @param string $errorQueueName
+     * @param string $routingKey
+     *
+     * @return \Generated\Shared\Transfer\RabbitMqOptionTransfer
+     */
+    protected function createErrorQueueBinding($errorQueueName, $routingKey)
```
Comment (ehsanmx) [inline src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: remove this method

**Comment by ehsanmx** [comment]
Also Travis is red, there is followup PR for travis can you merge it this one as well
https://github.com/spryker/rabbit-mq/pull/44/files

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php (hunk 1 above)
Comment (alexanderM91) [reply src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: In this case the best option to store RabbitMqConfig::getQueueOptions() on the project level

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php (hunk 3 above)
Comment (alexanderM91) [reply src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: removed

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php [hunk 4]
```diff
@@ -75,34 +106,85 @@ protected function getQueueConnectionConfigs(): array
...
     /**
-     * @return \Generated\Shared\Transfer\QueueConnectionTransfer
+     * @param string $queueName
+     * @param string $boundQueueName
+     * @param string $routingKey
+     *
+     * @return \Generated\Shared\Transfer\RabbitMqOptionTransfer
      */
-    public function getDefaultQueueConnectionConfig(): QueueConnectionTransfer
+    protected function createQueueOption($queueName, $boundQueueName, $routingKey = '')
```
[the remark this answers is not in the document]
Comment (ehsanmx) [reply src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: What will happen if this method never gets overwritten in the project?

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php [hunk 5]
```diff
@@ -75,34 +106,85 @@ protected function getQueueConnectionConfigs(): array
...
+     * @return \Generated\Shared\Transfer\RabbitMqOptionTransfer
+     */
+    protected function createQueueBinding($queueName)
+    {
+        $queueOptionTransfer = new RabbitMqOptionTransfer();
+        $queueOptionTransfer
+            ->setQueueName($queueName)
+            ->setDurable(true)
+            ->setNoWait(false)
+            ->addRoutingKey('');
```
Comment (ehsanmx) [inline src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: is this mandatory?

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (ehsanmx) [reply src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: `createExchangeOptionTransfer` is a better name

File: src/Spryker/Client/RabbitMq/RabbitMqConfig.php [hunk 6]
```diff
@@ -75,34 +106,85 @@ protected function getQueueConnectionConfigs(): array
...
+        return $queueOptionTransfer;
+    }
+
+    /**
+     * @param string $queueName
+     * @param string $routingKey
+     *
+     * @return \Generated\Shared\Transfer\RabbitMqOptionTransfer
+     */
+    protected function createQueueBinding($queueName)
```
Comment (ehsanmx) [inline src/Spryker/Client/RabbitMq/RabbitMqConfig.php]: `createQueueOptionTransfer` is a better name

File: .travis.yml [hunk 7]
```diff
@@ -17,6 +17,10 @@ matrix:
 
   fast_finish: true
 
+  allow_failures:
+      -   php: 7.2
+      -   php: 7.4snapshot
```
Comment (dereuromark) [inline .travis.yml]: This is not allowed.
Please fix this as per https://github.com/spryker/rabbit-mq/pull/44 requested last year with highest prio.

**Pull request merged**